### DISCUSSION TITLE
Urbanity Road Map

### DISCUSSION THREAD

**Discussion by winstonyym** [winstonyym/urbanity#2] [General]
# Urbanity - Roadmap

This document describes the current status and the upcoming milestones of the Urbanity project.

*Updated: Thursday, 25 May 2023 15:12:46 GMT*

#### Milestone Summary

| Status | Milestone | Labels | ETA |
| :---: | :--- | :---: | :---: |
| ✔ | **[Implement Mapping Interface](#implement-mapping-interface)** | `ready` | Aug 22 2022 |
| ✔ | **[Add City Scale Network Construction Module](#add-city-scale-network-construction-module)** |  `ready` |  Sep 08 2022 |
| ✔ | **[Compute Building Morphological Information for Network](#compute-building-morphological-information-for-network)** | `ready` | Sep 16 2022 |
| ✔ | **[Integrate High Resolution Urban Population Data](#integrate-high-resolution-urban-population-data)** |  `ready` |  Oct 11 2022 |
| ✔ | **[Develop SVI Module and Data Integration](#develop-svi-module-and-data-integration)** |  `ready` |  Oct 28 2022 |
| ✔ | **[Build Aggregate Urban Statistics Functionality](#build-aggregate-urban-statistics-functionality)** |  `ready` |  Nov 14 2022 |
| ✔ | **[Setup Package Documentation on Readthedocs](#setup-package-documentation-on-readthedocs)** | `ready` | Nov 18 2022 |
| ✔ | **[Develop Urbanity Dashboard](#develop-urbanity-dashboard)** |  `ready` |  Feb 10 2023 |
| ✔ | **[Establish Global Network Data Repository and Tiled Country Level Data](#establish-global-network-data-repository-and-tiled-country-level-data)** |  `ready` |  Mar 27 2023 |
| ✔ | **[Include Computation of Network Edge Attributes](#include-computation-of-network-edge-attributes)** |  `ready` |  Apr 12 2023 |
| 🚀 | **Implement Heterogeneous Graph Representation Learning \| Network-2-Building** |  `to-do` |  Jul 2023 |
| 🚀 | **Integrate High Resolution Satellite Features into Urban Networks \| Spatio-Temporal Learning** |  `to-do` |  Nov 2023 |
| 🚀 | **Integrate Contextual Features from 3D City Models into Urban Networks** |  `to-do` |  March 2024 |
| 🚀 | **Develop Graph Reinforcement Learning Use Cases for Node, Edge, and Graph Level Predictive Tasks \| City Scale Carbon Mapping \| GDP Prediction** |  `to-do` |  June 2024 |

<a name="implement-mapping-interface"></a>
#### Implement Mapping Interface

> Added features:
* Complete ipyleaflet map object construction
* Link country names to map coordinates for streamlined viewing
* Methods to add and remove popular vector files (.geojson/.shp)
* Handler tools to update base map layers, draw polygons and bounding boxes

This will allow users a visually intuitive interface to navigate their target area of interest.

<a name="add-city-scale-network-construction-module"></a>
#### Add City Scale Network Construction Module

> Added features:
* Users can automatically retrieve city-scale road networks from GeoFabrik
* Methods allow users to create various network representation (primal planar or dual networks)
* Network structure attributes are included in network

This will allow users a quick and easy way to retrieve network data from OpenStreetMap.

<a name="compute-building-morphological-information-for-network"></a>
#### Compute Building Morphological Information for Network

> Added features:
* Enable automatic integration of building morphology attributes into network nodes
* Enable automatic integration of urban points of interest into network nodes
* Modify function call to allow users to toggle on and off feature request.

This will allow users the flexibility to request building and points of interest data from OpenStreetMap.

<a name="integrate-high-resolution-urban-population-data"></a>
#### Integrate High Resolution Urban Population Data

> Added features:
* Methods to call and merge global urban population data 
* Merge population data for large countries
* Methods to preprocess both raster and vector population data

Users will gain access to Meta's High Resolution Population Density Maps through function API call and be able to request population subgroup information at 30m resolution.

<a name="develop-svi-module-and-data-integration"></a>
#### Develop SVI Module and Data Integration

> Added features:
* Screen and pre-process ~97 million Mapillary streetview images 
* Segment ~4 million images to obtain semantic indicators
* Release SVI data under open source repository

Provide quick access to SVI indicators for 50 cities around the world.

<a name="build-aggregate-urban-statistics-functionality"></a>
#### Build Aggregate Urban Statistics Functionality

> Added features:
* Add function to compute aggregate statistics for arbitrary urban zones
* Include API call request for network attributes
*  Include API call request for SVI attributes
*  Include API call request for building morphology attributes
*  Include API call request for POIs attributes
*  Include API call request for population attributes

Users can obtain aggregate urban statistics for subzones that do not correspond to census tract.

<a name="setup-package-documentation-on-readthedocs"></a>
#### Setup Package Documentation on Readthedocs

> Added features:
* Add detailed package module documentation
* Add example notebook to describe basic package functionality
* Add example notebook to demonstrate road network prediction
* Add example notebook to show network assortativity analysis

Provide documentation support and usage instruction for users. 

<a name="develop-urbanity-dashboard"></a>
#### Develop Urbanity Dashboard

> Added features:
* Support quick descriptive visualisation of Urbanity network indicators
* Facilitate quick comparative analyses between cities
* Allow low barrier exploratory analysis of urban networks

Release docker image containers for both MacOS and Window users.

<a name="establish-global-network-data-repository-and-tiled-country-level-data"></a>
#### Establish Global Network Data Repository and Tiled Country Level Data

> Added features:
* Pre-computed network data for 50 global cities
* Add city tracking page and city coverage request page
* Tiled population data to allow for rapid data retrieval of population information for any urban scale

Openly release feature-rich network dataset for 50 global cities to facilitate usage of network data for various use case and applications.

<a name="include-computation-of-network-edge-attributes"></a>
#### Include Computation of Network Edge Attributes

> Added features:
* Provide users access to edge level attributes
* Provide automated nearest edge interpolation and aggregation function 

Users can now retrieve contextual features not only for node but also network edges. Edges employ nearest feature interpolation to accumulate features. For example, POIs are assigned to their nearest network edge based on Euclidean distance.

**Comment by Ultios** [comment]
Nice roadmap, I'm tempted to do the same.
Anyway, is there any update on the Jul 23 and Nov 23 milestone?

**Reply by winstonyym** [reply]
Hi @Ultios, really grateful for your interest in our package. Yes, we are in the process to add functionalities to process multi-modal data and build heterogeneous graphs for cities. Progress is slower than the original timeline (due to conferences and other life commitments) but it should be out by the end of the year!
